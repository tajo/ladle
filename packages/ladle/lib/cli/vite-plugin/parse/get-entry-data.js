import fs from "fs";
import path from "path";
import debugFactory from "debug";
import { traverse } from "../babel.js";
import { getFileId } from "../naming-utils.js";
import getAst from "../get-ast.js";
import getDefaultExport from "./get-default-export.js";
import getStorynameAndMeta from "./get-storyname-and-meta.js";
import getNamedExports from "./get-named-exports.js";
import { IMPORT_ROOT } from "../utils.js";
import mdxToStories from "../mdx-to-stories.js";

const debug = debugFactory("ladle:vite");

/**
 * @param {string[]} entries
 */
export const getEntryData = async (entries) => {
  /**
   * @type {import('../../../shared/types').EntryData}
   */
  const entryData = {};
  for (let entry of entries) {
    debug(`Parsing ${entry}`);
    entryData[entry] = await getSingleEntry(entry);
  }
  return entryData;
};

/**
 * @param {string} entry
 */
export const getSingleEntry = async (entry) => {
  // fs.promises.readFile is much slower and we don't mind hogging
  // the whole CPU core since this is blocking everything else
  const fileCode = fs.readFileSync(path.join(IMPORT_ROOT, entry), "utf8");
  const code = entry.endsWith(".mdx")
    ? await mdxToStories(fileCode, entry, true)
    : fileCode;
  /** @type {import('../../../shared/types').ParsedStoriesResult} */
  const result = {
    entry,
    stories: [],
    exportDefaultProps: { title: undefined, meta: undefined },
    namedExportToMeta: {},
    namedExportToStoryName: {},
    storyParams: {},
    //@ts-ignore
    storySource: code.replace(/\r/g, ""),
    fileId: getFileId(entry),
  };
  //@ts-ignore
  const ast = getAst(code, entry);
  traverse(ast, {
    Program: getStorynameAndMeta.bind(this, result),
  });
  traverse(ast, {
    ExportDefaultDeclaration: getDefaultExport.bind(this, result),
  });
  traverse(ast, {
    ExportNamedDeclaration: getNamedExports.bind(this, result),
  });
  debug(`Parsed data for ${entry}:`);
  debug(result);
  return result;
};
