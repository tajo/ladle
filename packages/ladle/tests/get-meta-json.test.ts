import { test, expect } from "vitest";
import { getEntryData } from "../lib/cli/vite-plugin/parse/get-entry-data.js";
import {
  getMetaJsonString,
  getMetaJson,
} from "../lib/cli/vite-plugin/generate/get-meta-json.js";

test("Single file with two stories", async () => {
  const entryData = await getEntryData(["tests/fixtures/animals.stories.tsx"]);
  const list = getMetaJsonString(entryData);
  expect(list).toMatchSnapshot();
});

test("Capital letters in story names converted into delimiters", async () => {
  const entryData = await getEntryData([
    "tests/fixtures/capitalization.stories.tsx",
  ]);
  const list = getMetaJsonString(entryData);
  expect(list).toMatchSnapshot();
});

test("Capital letters in the filename converted into delimiters", async () => {
  const entryData = await getEntryData([
    "tests/fixtures/filenameCapitalization.stories.tsx",
  ]);
  const list = getMetaJsonString(entryData);
  expect(list).toMatchSnapshot();
});

test("Turn file name delimiters into spaces and levels correctly", async () => {
  const entryData = await getEntryData([
    "tests/fixtures/our-animals--mammals.stories.tsx",
  ]);
  const list = getMetaJsonString(entryData);
  expect(list).toMatchSnapshot();
});

test("Default title is used instead of the file name", async () => {
  const entryData = await getEntryData([
    "tests/fixtures/default-title.stories.tsx",
  ]);
  const list = getMetaJsonString(entryData);
  expect(list).toMatchSnapshot();
});

test("Story name replaces named export as a story name", async () => {
  const entryData = await getEntryData([
    "tests/fixtures/storyname.stories.tsx",
  ]);
  const list = getMetaJsonString(entryData);
  expect(list).toMatchSnapshot();
});

test("Extract default meta", async () => {
  const entryData = await getEntryData([
    "tests/fixtures/default-meta.stories.tsx",
  ]);
  const list = getMetaJsonString(entryData);
  expect(list).toMatchSnapshot();
});

test("Extract and merge story meta", async () => {
  const entryData = await getEntryData([
    "tests/fixtures/story-meta.stories.tsx",
  ]);
  const list = getMetaJsonString(entryData);
  expect(list).toMatchSnapshot();
});

test("Test multiple stories", async () => {
  const entryData = await getEntryData([
    "tests/fixtures/storyname.stories.tsx",
    "tests/fixtures/story-meta.stories.tsx",
  ]);
  const list = getMetaJson(entryData);
  expect(list).toMatchSnapshot();
});
