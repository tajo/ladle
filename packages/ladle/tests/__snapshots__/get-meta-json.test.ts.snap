// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Capital letters in story names converted into delimiters 1`] = `
"{
  \\"about\\": {
    \\"homepage\\": \\"https://www.ladle.dev\\",
    \\"github\\": \\"https://github.com/tajo/ladle\\",
    \\"version\\": 1
  },
  \\"stories\\": {
    \\"capitalization--big-barking-dog\\": {
      \\"name\\": \\"Big barking dog\\",
      \\"levels\\": [
        \\"Capitalization\\"
      ],
      \\"locStart\\": 8,
      \\"locEnd\\": 10,
      \\"filePath\\": \\"tests/fixtures/capitalization.stories.tsx\\",
      \\"namedExport\\": \\"BigBarkingDog\\",
      \\"meta\\": {}
    },
    \\"capitalization--blue-tiny-cat\\": {
      \\"name\\": \\"Blue tiny cat\\",
      \\"levels\\": [
        \\"Capitalization\\"
      ],
      \\"locStart\\": 4,
      \\"locEnd\\": 6,
      \\"filePath\\": \\"tests/fixtures/capitalization.stories.tsx\\",
      \\"namedExport\\": \\"BlueTinyCat\\",
      \\"meta\\": {}
    }
  }
}"
`;

exports[`Capital letters in the filename converted into delimiters 1`] = `
"{
  \\"about\\": {
    \\"homepage\\": \\"https://www.ladle.dev\\",
    \\"github\\": \\"https://github.com/tajo/ladle\\",
    \\"version\\": 1
  },
  \\"stories\\": {
    \\"filename-capitalization--test\\": {
      \\"name\\": \\"Test\\",
      \\"levels\\": [
        \\"Filename capitalization\\"
      ],
      \\"locStart\\": 4,
      \\"locEnd\\": 6,
      \\"filePath\\": \\"tests/fixtures/filenameCapitalization.stories.tsx\\",
      \\"namedExport\\": \\"Test\\",
      \\"meta\\": {}
    }
  }
}"
`;

exports[`Default title is used instead of the file name 1`] = `
"{
  \\"about\\": {
    \\"homepage\\": \\"https://www.ladle.dev\\",
    \\"github\\": \\"https://github.com/tajo/ladle\\",
    \\"version\\": 1
  },
  \\"stories\\": {
    \\"title--cat\\": {
      \\"name\\": \\"Cat\\",
      \\"levels\\": [
        \\"Title\\"
      ],
      \\"locStart\\": 8,
      \\"locEnd\\": 10,
      \\"filePath\\": \\"tests/fixtures/default-title.stories.tsx\\",
      \\"namedExport\\": \\"Cat\\",
      \\"meta\\": {}
    }
  }
}"
`;

exports[`Extract and merge story meta 1`] = `
"{
  \\"about\\": {
    \\"homepage\\": \\"https://www.ladle.dev\\",
    \\"github\\": \\"https://github.com/tajo/ladle\\",
    \\"version\\": 1
  },
  \\"stories\\": {
    \\"title--cat\\": {
      \\"name\\": \\"Cat\\",
      \\"levels\\": [
        \\"Title\\"
      ],
      \\"locStart\\": 14,
      \\"locEnd\\": 16,
      \\"filePath\\": \\"tests/fixtures/story-meta.stories.tsx\\",
      \\"namedExport\\": \\"Cat\\",
      \\"meta\\": {
        \\"baseweb\\": {
          \\"theme\\": \\"dark\\",
          \\"browsers\\": [
            \\"chrome\\",
            \\"firefox\\"
          ],
          \\"width\\": \\"500px\\"
        },
        \\"links\\": true
      }
    },
    \\"title--dog\\": {
      \\"name\\": \\"Dog\\",
      \\"levels\\": [
        \\"Title\\"
      ],
      \\"locStart\\": 26,
      \\"locEnd\\": 28,
      \\"filePath\\": \\"tests/fixtures/story-meta.stories.tsx\\",
      \\"namedExport\\": \\"Dog\\",
      \\"meta\\": {
        \\"baseweb\\": {
          \\"theme\\": \\"dark\\",
          \\"browsers\\": [
            \\"chrome\\",
            \\"webkit\\"
          ]
        }
      }
    }
  }
}"
`;

exports[`Extract default meta 1`] = `
"{
  \\"about\\": {
    \\"homepage\\": \\"https://www.ladle.dev\\",
    \\"github\\": \\"https://github.com/tajo/ladle\\",
    \\"version\\": 1
  },
  \\"stories\\": {
    \\"title--cat\\": {
      \\"name\\": \\"Cat\\",
      \\"levels\\": [
        \\"Title\\"
      ],
      \\"locStart\\": 13,
      \\"locEnd\\": 15,
      \\"filePath\\": \\"tests/fixtures/default-meta.stories.tsx\\",
      \\"namedExport\\": \\"Cat\\",
      \\"meta\\": {
        \\"baseweb\\": {
          \\"foo\\": \\"title\\"
        }
      }
    }
  }
}"
`;

exports[`Single file with two stories 1`] = `
"{
  \\"about\\": {
    \\"homepage\\": \\"https://www.ladle.dev\\",
    \\"github\\": \\"https://github.com/tajo/ladle\\",
    \\"version\\": 1
  },
  \\"stories\\": {
    \\"animals--cat\\": {
      \\"name\\": \\"Cat\\",
      \\"levels\\": [
        \\"Animals\\"
      ],
      \\"locStart\\": 4,
      \\"locEnd\\": 6,
      \\"filePath\\": \\"tests/fixtures/animals.stories.tsx\\",
      \\"namedExport\\": \\"Cat\\",
      \\"meta\\": {}
    },
    \\"animals--dog\\": {
      \\"name\\": \\"Dog\\",
      \\"levels\\": [
        \\"Animals\\"
      ],
      \\"locStart\\": 8,
      \\"locEnd\\": 10,
      \\"filePath\\": \\"tests/fixtures/animals.stories.tsx\\",
      \\"namedExport\\": \\"Dog\\",
      \\"meta\\": {}
    }
  }
}"
`;

exports[`Story name replaces named export as a story name 1`] = `
"{
  \\"about\\": {
    \\"homepage\\": \\"https://www.ladle.dev\\",
    \\"github\\": \\"https://github.com/tajo/ladle\\",
    \\"version\\": 1
  },
  \\"stories\\": {
    \\"storyname--capital-city\\": {
      \\"name\\": \\"Capital city\\",
      \\"levels\\": [
        \\"Storyname\\"
      ],
      \\"locStart\\": 15,
      \\"locEnd\\": 17,
      \\"filePath\\": \\"tests/fixtures/storyname.stories.tsx\\",
      \\"namedExport\\": \\"CapitalCity\\",
      \\"meta\\": {}
    },
    \\"storyname--champs-élysées\\": {
      \\"name\\": \\"Champs élysées\\",
      \\"levels\\": [
        \\"Storyname\\"
      ],
      \\"locStart\\": 19,
      \\"locEnd\\": 21,
      \\"filePath\\": \\"tests/fixtures/storyname.stories.tsx\\",
      \\"namedExport\\": \\"CapitalReplaced\\",
      \\"meta\\": {}
    },
    \\"storyname--doggo\\": {
      \\"name\\": \\"Doggo\\",
      \\"levels\\": [
        \\"Storyname\\"
      ],
      \\"locStart\\": 4,
      \\"locEnd\\": 9,
      \\"filePath\\": \\"tests/fixtures/storyname.stories.tsx\\",
      \\"namedExport\\": \\"Cat\\",
      \\"meta\\": {}
    }
  }
}"
`;

exports[`Test multiple stories 1`] = `
{
  "about": {
    "github": "https://github.com/tajo/ladle",
    "homepage": "https://www.ladle.dev",
    "version": 1,
  },
  "stories": {
    "storyname--capital-city": {
      "filePath": "tests/fixtures/storyname.stories.tsx",
      "levels": [
        "Storyname",
      ],
      "locEnd": 17,
      "locStart": 15,
      "meta": {},
      "name": "Capital city",
      "namedExport": "CapitalCity",
    },
    "storyname--champs-élysées": {
      "filePath": "tests/fixtures/storyname.stories.tsx",
      "levels": [
        "Storyname",
      ],
      "locEnd": 21,
      "locStart": 19,
      "meta": {},
      "name": "Champs élysées",
      "namedExport": "CapitalReplaced",
    },
    "storyname--doggo": {
      "filePath": "tests/fixtures/storyname.stories.tsx",
      "levels": [
        "Storyname",
      ],
      "locEnd": 9,
      "locStart": 4,
      "meta": {},
      "name": "Doggo",
      "namedExport": "Cat",
    },
    "title--cat": {
      "filePath": "tests/fixtures/story-meta.stories.tsx",
      "levels": [
        "Title",
      ],
      "locEnd": 16,
      "locStart": 14,
      "meta": {
        "baseweb": {
          "browsers": [
            "chrome",
            "firefox",
          ],
          "theme": "dark",
          "width": "500px",
        },
        "links": true,
      },
      "name": "Cat",
      "namedExport": "Cat",
    },
    "title--dog": {
      "filePath": "tests/fixtures/story-meta.stories.tsx",
      "levels": [
        "Title",
      ],
      "locEnd": 28,
      "locStart": 26,
      "meta": {
        "baseweb": {
          "browsers": [
            "chrome",
            "webkit",
          ],
          "theme": "dark",
        },
      },
      "name": "Dog",
      "namedExport": "Dog",
    },
  },
}
`;

exports[`Turn file name delimiters into spaces and levels correctly 1`] = `
"{
  \\"about\\": {
    \\"homepage\\": \\"https://www.ladle.dev\\",
    \\"github\\": \\"https://github.com/tajo/ladle\\",
    \\"version\\": 1
  },
  \\"stories\\": {
    \\"our-animals--mammals--cat\\": {
      \\"name\\": \\"Cat\\",
      \\"levels\\": [
        \\"Our animals\\",
        \\"Mammals\\"
      ],
      \\"locStart\\": 4,
      \\"locEnd\\": 6,
      \\"filePath\\": \\"tests/fixtures/our-animals--mammals.stories.tsx\\",
      \\"namedExport\\": \\"Cat\\",
      \\"meta\\": {}
    }
  }
}"
`;
